commit a5fbe28a1c3c397681e3e90eefc177b552eafe18 (refs/changes/99/99/4)
Author: Pedro P. Candel <pedro@joyent.com>
Date:   2016-07-21T13:40:43+02:00 (3 years, 3 months ago)
    
    TOOLS-1487 sdcadm improve function for handling circular references
    Reviewed by: Trent Mick <trent.mick@joyent.com>
    Reviewed by: Joshua M. Clulow <jmc@joyent.com>

diff --git a/lib/history.js b/lib/history.js
index ae03cac..ae0df76 100644
--- a/lib/history.js
+++ b/lib/history.js
@@ -30,17 +30,12 @@ var UUID_RE = /^[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}$/;
 
 // Intended to be used as second argument to JSON.Stringify() in order to
 // prevent 'TypeError: Converting circular structure to JSON'
-function replacer() {
-    var i = 0;
+function safeCycles() {
     var objs = [];
     return function (key, value) {
-        // Avoid "RangeError: Maximum call stack size exceeded"
-        // (Note this is an arbitrary number, need to set max depth
-        // at some point)
-        if (i >= 29) {
-            return '[Unknown]';
+        if (!value || typeof (value) !== 'object') {
+            return value;
         }
-        i += 1;
         if (objs.indexOf(value) !== -1) {
             return '[Circular]';
         } else {
@@ -54,7 +49,7 @@ function replacer() {
 function saveHistoryToFile(fname, history, cb) {
     var s;
     try {
-        s = JSON.stringify(history, replacer());
+        s = JSON.stringify(history, safeCycles());
     } catch (e) {
         s = '{}';
     }
